### PR TITLE
Signal spectral window default fix

### PR DESCRIPTION
**Pull request by NickleDave**
Resolves issue #6817 (hopefully). The _spectral_helper function now checks whether window is array_like and if so changes nperseg to the length of that array. Otherwise, if nperseg was not specified, then it is set to the default of 256. This means that that all functions with an nperseg argument that call _spectral_helper now have the default nperseg=None (i.e. not specified) in their argument lists instead of nperseg=256.

Tested with example code from the docs and everything looks the same. Examples also look as I would expect when using the default 256 instead of nperseg=1024 in the examples.

### CODE REVIEW

File: scipy/signal/spectral.py [hunk 1]
```diff
@@ -507,6 +507,11 @@ def spectrogram(x, fs=1.0, window=('tukey',.25), nperseg=256, noverlap=None,
     >>> plt.xlabel('Time [sec]')
     >>> plt.show()
     """
+    #parse window; if array like, then set nperseg = win.shape
+    if isinstance(window, string_types) or type(window) is tuple:
+        if nperseg == None: # if nperseg not specified by user or calling function
```
Comment (Eric89GXL) [inline scipy/signal/spectral.py]: `if nperseg is None`

File: scipy/signal/spectral.py [hunk 2]
```diff
@@ -507,6 +507,11 @@ def spectrogram(x, fs=1.0, window=('tukey',.25), nperseg=256, noverlap=None,
     >>> plt.xlabel('Time [sec]')
     >>> plt.show()
     """
+    #parse window; if array like, then set nperseg = win.shape
+    if isinstance(window, string_types) or type(window) is tuple:
```
Comment (Eric89GXL) [inline scipy/signal/spectral.py]: `or isinstance(window, tuple)`

File: scipy/signal/spectral.py [hunk 3]
```diff
@@ -507,6 +507,11 @@ def spectrogram(x, fs=1.0, window=('tukey',.25), nperseg=256, noverlap=None,
...
+        if nperseg == None: # if nperseg not specified by user or calling function
+            nperseg = 256 # change to default
+
```
Comment (Eric89GXL) [inline scipy/signal/spectral.py]: I think the logic here is incomplete. I can call for example `spectrogram(..., nperseg=None, window=np.ones(100), noverlap=None)`. A couple lines below there is a `if noverlap is None: noverlap = nperseg // 8`, but based on your changes `nperseg` will still be `None`. So something needs to change, probably an `elif nperseg is None` to pair with your `if isinstance(...)` line.

But really the cleaner thing might be to move the window triaging up, and nest the `nperseg is None` checks in there.

File: scipy/signal/spectral.py [hunk 4]
```diff
@@ -766,6 +771,18 @@ def _spectral_helper(x, y, fs=1.0, window='hann', nperseg=256,
                 pad_shape[-1] = x.shape[-1] - y.shape[-1]
                 y = np.concatenate((y, np.zeros(pad_shape)), -1)
 
+    #parse window; if array like, then set nperseg = win.shape
+    if isinstance(window, string_types) or type(window) is tuple:
```
Comment (Eric89GXL) [inline scipy/signal/spectral.py]: same comments about `ininstance` and `nperseg is None`.

Any way these checks could be refactored into a private function instead of repeated multiple places?

Comment (Eric89GXL) [inline scipy/signal/spectral.py]: I'm thinking `win, nperseg, noverlap = _triage_segments(window, nperseg, noverlap)` in both places could work.

**Review by larsoner — COMMENTED** [review]
Also, you should add some tests for your `window=ndarray` case. They would have caught the `spectrogram` logic complaint that I mentioned.

**Comment by Eric89GXL** [comment]
It looks like these changes break some existing tests, too:

https://travis-ci.org/scipy/scipy/jobs/180632956#L1444

**Comment by NickleDave** [comment]
@Eric89GXL I made those changes.
Not sure how many tests to add. In test_spectral.py there's test_window_external (e.g. line 352 for Welch) but it specifies nperseg.
Should I write a separate test with nperseg=None? And another where nperseg is not specified? Or just group them all under that test?
Also would it make sense to throw a warning for the edge case:
```python
win = signal.get_window('hann',8)
f,p = welch(x, 10, win, nperseg=16)
```
where a user specifies window=ndarray with length x but also sets nperseg to some different value y, .e.g. "Warning: nperseg was set to length of window x". If so and there's other functions with similar warnings you can point me to, I'll try to copy their format

**Comment by Eric89GXL** [comment]
> In test_spectral.py there's test_window_external (e.g. line 352 for
Welch) but it specifies nperseg. Should I write a separate test with
nperseg=None? And another where nperseg is not specified? Or just group
them all under that test?

It should be okay to group them all in the same test or split them, up to
you.

> where a user specifies window=ndarray with length x but also sets nperseg
to some different value y

I would throw an error if they specify nperseg and their window is a
different size. Please add a test (assert_raises) for this, too.

Let me know when you make these changes and the CIs are happy and I'll look
at the code again.

**Comment by NickleDave** [comment]
@Eric89GXL this now passes all checks except travis-ci for 3.5, but it's not at all clear to me what's going wrong with the 3.5

**Comment by Eric89GXL** [comment]
You mean this traceback?
```
======================================================================
ERROR: test_window_correction (test_spectral.TestWelch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/tests/test_spectral.py", line 488, in test_window_correction
    scaling='spectrum')
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/spectral.py", line 273, in welch
    return_onesided, scaling, axis)
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/spectral.py", line 391, in csd
    mode='psd')
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/spectral.py", line 774, in _spectral_helper
    win, nperseg = _triage_segments(window, nperseg)
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/spectral.py", line 979, in _triage_segments
    win = get_window(window, nperseg)
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/windows.py", line 1768, in get_window
    return winfunc(*params)
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/windows.py", line 812, in hann
    w = _cos_win(M, [0.5, 0.5])
  File "/home/travis/virtualenv/python3.5.2/lib/python3.5/site-packages/scipy/signal/windows.py", line 114, in _cos_win
    w = np.zeros(M)
TypeError: 'float' object cannot be interpreted as an integer
```
Somehow `M` must be a `float` at this point, but needs to be an integer. Maybe 3.5 has stricter checks for this sort of thing.

**Comment by NickleDave** [comment]
@Eric89GXL fixed that, thank you. 

```testWelch.test_window_correction``` called ```welch``` with ```nperseg``` set to a float.
```nperseg``` becomes ```M``` when passed to ```winfunc``` at the bottom of the stack trace.
I just wrapped the line that calculates ```nperseg``` in an ```int()``` since obviously the point of the test is not to check that ```np.zeros(M)``` will give an error when passed a non-integer value for M.
So now it's
```Python
        fs = 1e4
        nperseg = int(fs//10)
#...
            _, p_spec = welch(x, fs=fs, nperseg=nperseg, window=window,
                              scaling='spectrum')
```
and that works fine

Not related, but assuming this code passes muster, can I add myself to THANKS.txt?
Thank you for walking me through this--if it wasn't obvious already, this is the first time I've made an actual code contribution

File: scipy/signal/spectral.py [hunk 5]
```diff
@@ -952,3 +951,40 @@ def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft):
     result = fftpack.fft(result, n=nfft)
 
     return result
+
+def _triage_segments(window, nperseg):
+    """
+    private function that handles defaults.
+    called by _spectral_helper, and also by spectrogram because it requires a
```
Comment (Eric89GXL) [inline scipy/signal/spectral.py]: Clean up the definition line, and add proper capitalization, etc. (see how other helpers are formatted)

File: scipy/signal/tests/test_spectral.py [hunk 6]
```diff
@@ -474,7 +476,7 @@ def test_padded_freqs(self):
     def test_window_correction(self):
         A = 20
         fs = 1e4
-        nperseg = fs//10
+        nperseg = int(fs//10)
```
Comment (Eric89GXL) [inline scipy/signal/tests/test_spectral.py]: you can just do `int(fs) // 10` because the `//` is integer division

File: scipy/signal/tests/test_spectral.py [hunk 7]
```diff
@@ -660,6 +662,8 @@ def test_window_external(self):
         fe, pe = csd(x, x, 10, win, 8)
         assert_array_almost_equal_nulp(p, pe)
         assert_array_almost_equal_nulp(f, fe)
+        assert_raises(ValueError, csd, x, x,
```
Comment (Eric89GXL) [inline scipy/signal/tests/test_spectral.py]: add a comment about what is invalid here, e.g. `# nperseg != len(win)`

**Review by larsoner — COMMENTED** [review]
Otherwise LGTM. Thanks for working on this bit.

**Comment by Eric89GXL** [comment]
@e-q do you have time to look, too?

File: scipy/signal/spectral.py [hunk 8]
```diff
@@ -141,7 +141,7 @@ def periodogram(x, fs=1.0, window=None, nfft=None, detrend='constant',
                  scaling, axis)
 
 
-def welch(x, fs=1.0, window='hann', nperseg=256, noverlap=None, nfft=None,
+def welch(x, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
```
Comment (e-q) [inline scipy/signal/spectral.py]: The changes in the default values need to be reflected in the docstrings. For example, `noverlap` also defaults to `None` and the docstring states: "If None, `noverlap = nperseg // 2`.  Defaults to None."

File: scipy/signal/spectral.py [hunk 9]
```diff
@@ -952,3 +951,40 @@ def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft):
...
+    else:
+        win = np.asarray(window)
+        if len(win.shape) != 1:
+            raise ValueError('window must be 1-D')
+        if nperseg is None:
+            nperseg = win.shape[0]
+        elif nperseg is not None:
+            if nperseg != win.shape[0]:
+                raise ValueError("Value specified for nperseg is different from"
+                                 " length of ndarray provided as window.")
```
Comment (e-q) [inline scipy/signal/spectral.py]: The window does not, strictly speaking, need to be provided as an `ndarrary`. It's probably clearer to say something like "window must have length of nperseg."

File: scipy/signal/spectral.py [hunk 10]
```diff
@@ -766,11 +770,15 @@ def _spectral_helper(x, y, fs=1.0, window='hann', nperseg=256,
...
 
+    # parse window; if array like, then set nperseg = win.shape
+    win, nperseg = _triage_segments(window, nperseg)              
+        
     # X and Y are same length now, can test nperseg with either
     if x.shape[-1] < nperseg:
         warnings.warn('nperseg = {0:d}, is greater than input length = {1:d}, '
                       'using nperseg = {1:d}'.format(nperseg, x.shape[-1]))
         nperseg = x.shape[-1]
+        win, nperseg = _triage_segments(window, nperseg)
```
Comment (e-q) [inline scipy/signal/spectral.py]: It looks like this would not catch the case where the window's length is greater than the signal length. 

If `nperseg=None` and the window has greater length than the signal, we should error out, since there's no good way to adapt the window to the signal. There should be test for this case as well.

File: scipy/signal/tests/test_spectral.py [hunk 11]
```diff
@@ -122,7 +122,7 @@ def test_window_external(self):
         fe, pe = periodogram(x, 10, win)
         assert_array_almost_equal_nulp(p, pe)
         assert_array_almost_equal_nulp(f, fe)
-
+
```
Comment (e-q) [inline scipy/signal/tests/test_spectral.py]: Looks like unnecessary whitespace. Weird that the pep8/pyflakes check didn't catch this.

**Review by e-q — CHANGES_REQUESTED** [review]
Left some feedback for some missing pieces, thanks for working on this!

One more test that should be added is checking that the window length is really being used as the segment length when `nperseg` is given as `None`. This could be done by looking at the length of the return arrays, which for one-sided spectra should be `nperseg//2+1`.

File: scipy/signal/spectral.py (hunk 8 above)
Comment (NickleDave) [inline scipy/signal/spectral.py]: @e-q all functions still default to 256 where stated in the docstrings. It happens inside ```_triage_segments```.

I.e., as I have modified it, when someone calls any of these functions (except periodogram) one of three things happens:
1. they supply a string for window, e.g., 'hann', and then if nperseg==None, it is set to 256
2. they supply an ndarray, and if nperseg==None, it is set to the length of the ndarray
3. they supply an ndarray, and also a value for nperseg; if nperseg==len(ndarray), the expected happens, and if nperseg != len(ndarray), an error is thrown

That's the logic I came up with to resolve the original issue. I realize it's confusing that the docstring says one thing but the defaults in the keyword list say another.

**Comment by NickleDave** [comment]
@e-q about a test for whether window length is really being used as the segment length when nperseg is given as ```None```: this is checked by ```test_window_external``` in ```periodogram```, ```welch```, and ```csd```. ```coherence``` calls ```welch``` so there's no need for a separate check there. ```spectrogram``` does not have that test but I could add it.

```test_window_external``` does not explicitly set ```nperseg``` to ```None``` but isn't that something the user wants, not having to explicitly set nperseg to None when they pass an ndarray as the window? That's part of what makes it convenient to have it set automatically to the length of the window.

File: scipy/signal/spectral.py (hunk 10 above)
Comment (NickleDave) [inline scipy/signal/spectral.py]: I'm not quite clear what you mean by "when ```nperseg==None``` and the window has greater length than the signal".
As currently written, ```_spectral_helper``` checks whether nperseg is greater than the length of the signal, gives a warning, and then sets nperseg to the length of the window. This is in lines 780-784. In line 785 I added another call to ```_triage_segments``` so that it gets a new window corresponding to the new value of nperseg.

There's several tests in test_spectral that seem to be verifying whether this happens, e..g., ```TestCSD.test_short_data``` so I figured I should stay consistent with that behavior.

File: scipy/signal/tests/test_spectral.py (hunk 11 above)
Comment (NickleDave) [inline scipy/signal/tests/test_spectral.py]: Removed, thank you

File: scipy/signal/spectral.py (hunk 9 above)
Comment (NickleDave) [inline scipy/signal/spectral.py]: Changed that language

**Comment by NickleDave** [comment]
@Eric89GXL @e-q added another commit that addresses minor things but I'm a little confused about how to deal with other comments, please see my responses above.

**Comment by NickleDave** [comment]
Also I'm not clear why the Travis-CI check didn't run on the last commit. Something I did wrong?

**Comment by Eric89GXL** [comment]
It didn't run because there are conflicts, can you rebase?

File: scipy/signal/spectral.py (hunk 8 above)
Comment (e-q) [inline scipy/signal/spectral.py]: What I mean is that the docstring should reflect the true default in the function definition, and go on to explain the behavior. Explicit is better than implicit.

**Comment by e-q** [comment]
@NickleDave: It looks like `test_window_external` is setting `nperseg` explicitly as well as providing an external window. I think that this test should be modified to explicitly pass `nperseg=None` with the external window for the sake of clarity, and as a check that this new behavior does not break in the future if someone decides to make further changes to these functions and/or their defaults. 

The way I think about writing tests is less about replicating what users may do and more about providing checks that the code behaves consistently and as expected as the codebase evolves.

**Comment by NickleDave** [comment]
@e-q @Eric89GXL rebased, made tests for ```nperseg=None``` case explicit, rewrote docstrings so they reflect default and so that ```_triage_segments``` follows correct format.

Still not sure what to do about the case where the length of the ```window``` is greater than the length of the signal. Currently, ```_spectral_helper``` checks whether ```nperseg``` (*not window!*) is greater than the length of the signal, and if so gives a warning and sets ```nperseg``` to the signal length. I added a second call to ```_triage_segments``` so that ```win``` will have length equal to the new ```nperseg```.

There's several tests in test_spectral that seem to be verifying whether this happens, e..g., TestCSD.test_short_data so I figured I should stay consistent with that behavior.
Having it error out instead when ```window``` > signal length would be different behavior that might break existing code, right?

**Comment by rgommers** [comment]
@NickleDave something went wrong with your rebase, you have ~50 commits too many here. If you don't know how to fix that, please let us know.

**Comment by NickleDave** [comment]
@rgommers can you tell me the way that you would do the rebase? I thought I had figured it out but I don't want to make things worse.

**Comment by rgommers** [comment]
Looking at the history here, it seems like you have one series of commits based off of master, starting at commit 1da5916 and ending at commit 14b8ac7. Then you tried some mix of rebasing and merging which managed to duplicate that set of commits as well as pull in other commits that aren't related to this PR. Is that right?

If so, I'd try:

    git fetch upstream
    git reset --hard 14b8ac7
    git rebase upstream/master

That should work without any conflicts (I tried). If that looks good, what would also be nice if you then did an interactive rebase (``git rebase -i upstream/master``) to remove the pairs of commits where the second is a revert of the first commit.

Finally, for bonus points, you may want to look at our standard way of formatting commit messages: http://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_workflow.html#writing-the-commit-message. You can use ``reword`` in interactive rebase to do that (if that feels to risky, feel free to ignore this but keep it in mind for future commits)

**Comment by NickleDave** [comment]
@rgommers I did as you suggested but I'm not sure what to do next.
The changes don't show in this remote, only in my local copy.
Don't I need to do something like "git push -f origin:name-of-this-branch"?

Also:
1) I should have reset to the commit **before** the first failed rebase, right? I.e. hash e2837ff not 14b8ac7. 
2) Should I have done ```git rebase upstream/master``` on ```master```, or should I have done it on this branch? I first followed your directions exactly but was on this branch, then based on what I read online thought I should've done the rebase on master instead.

So as of now I have rebased master to upstream/master, and then reset --hard to e2837ff

**Comment by Eric89GXL** [comment]
> Don't I need to do something like "git push -f origin:name-of-this-branch"?

Yes, probably. You can double check with `git log` that you have one or two commits (however many you mean to add) on top of the latest one from `upstream/master`. If it all looks good you can push force.

> Should I have done git rebase upstream/master on master, or should I have done it on this branch?

You'd want to rebase whatever branch you're trying to fix, which in this case is this branch. So it sounds like you did the right thing.

> So as of now I have rebased master to upstream/master, and then reset --hard to e2837ff

A `reset --hard` will wipe out any rebasing you did. So hopefully you did a `reset --hard` first, and then did a `rebase`, in that order, as suggested by @rgommers.

**Comment by NickleDave** [comment]
@Eric89GXL @e-q @rgommers okay, I think I git it now :D (sorry)
Could you let me know your thoughts about whether functions should still give a warning and change ```nperseg``` to the length of the signal in the case where ```window``` is longer than the signal? @e-q seemed to be saying that this should instead result in an error. See comment above where I first tried to rebase

**Comment by Eric89GXL** [comment]
+1 for error if window length does not match nperseg

File: THANKS.txt [hunk 12]
```diff
@@ -166,7 +166,9 @@ Antonio Ribeiro for implementing irrnotch and iirpeak functions.
 Ilhan Polat for bug fixes on Riccati solvers.
 Sebastiano Vigna for code in the stats package related to Kendall's tau.
 John Draper for bug fixes.
+<<<<<<< 5be9d6205238a139be0b44a833617af8373f00db
```
Comment (rgommers) [inline THANKS.txt]: leftover from merge conflict

File: scipy/signal/spectral.py (hunk 8 above)
Comment (rgommers) [inline scipy/signal/spectral.py]: This comment is addressed, explanation of ``nperseg`` below LGTM

**Comment by rgommers** [comment]
Commit history looks as it should, nice. 

+1 for exception too.

**Comment by NickleDave** [comment]
Not sure if we're talking about the same errors...

It already raises an error when window length does not match ```nperseg```, in line 1018.

@e-q said above it should also raise an error when the window length is longer than the signal. Specifically he said this should happen in the case where ```nperseg=None```, i.e. when the user passes an ndarray for a window.

There was no such check before this pull request, but there was a check for whether ```nperseg``` was greater than the length of the signal.

That did *not* raise an error; instead ```_spectral_helper``` issued a warning and then changed ```nperseg``` to the length of the signal. This is in lines 792-797. (I had to add another call to ```_triage_segments``` to maintain this behavior)

In my last two commits, I added an error for the case @e-q specified, as well as tests for that case.
But as I've written it, the same error will also be raised when the user passes a tuple for the window and sets ```nperseg``` to some value greater than the length of the signal.
That is different from what the behavior used to be, where ```_spectral_helper``` issued a warning and set ```nperseg``` to the length of the signal.

So should I just axe that warning and let the functions always issue an error when the window is longer than the signal? That makes sense to me, in that it makes the user actively decide what to do instead of arbitrarily changing the window size to the signal length.

**Comment by Eric89GXL** [comment]
> So should I just axe that warning and let the functions always issue an error when the window is longer than the signal? That makes sense to me, in that it makes the user actively decide what to do instead of arbitrarily changing the window size to the signal length.

It does seem better to be explicit. If we raise an error, the fix for users is essentially `func(x, ..., window=my_window[:len(x)])` right? This seems acceptable, and likely to prevent more (hidden) problems than it might (explicitly) create.

**Comment by NickleDave** [comment]
Ok, so I removed the warning and rewrote the tests.

One of them, ```test_short_data``` explicitly tested for the ```nperseg``` warning and made sure that it occurred when nperseg was set to the default of 256. So it was testing for the previous behavior--change ```nperseg``` to the length of the input signal--and then made sure this gave the same result as when nperseg was set by the user to the length of the input.

It now tests that this raise an error instead.

I wrote a separate error message into ```_triage_segments``` that clearly states the error occurred because nperseg was set to the default, and also informs the user they how they should change nperseg to fix the error.

I should probably remove ```import warning``` since it's no longer used in spectral.py.

Please let me know if there's anything else I need to do for this PR.

**Comment by e-q** [comment]
> the fix for users is essentially func(x, ..., window=my_window[:len(x)]) right?

This isn't what users would want to do, as a truncated window will probably have some terrible spectral leakage. It would be something more like `window=my_window_function(N=len(x))`

**Comment by e-q** [comment]
I've been away, and gotten lost in the discussion about error vs. warning cases, so I wanted to summarize for myself some of the behaviors that are changing. I checked out this PR and ran a few cases.

- Unspecified `nperseg`, array-like `window` of length not equal to 256.
    - Current behavior: `ValueError` on window length not being `nperseg`
    - This PR: Set `nperseg` to length of `window`. 
    - This is the main fix of this PR 👍🏽

- Unspecified `nperseg`, array-like `window` longer than input signal
    - Current behavior: `ValueError` on window length not being `nperseg`
    - This PR: `ValueError` on window length exceeding signal length 
    - This is a more clear error message, and must be a `ValueError`, since there isn't a foolproof procedure to adjust an array-like window to a different arbitrary length. 👍🏽

- User-specified `nperseg` longer than input signal, array-like `window` matching too-long `nperseg`
    - Current behavior: Coerce `nperseg` to input signal length, warn user, `ValueError` on window length not matching `nperseg`.
    - This PR: `ValueError` on window length exceeding input signal length
    - Clearer error message 👍🏽

- Unspecified `nperseg`, string-like `window`, input signal length shorter than 256
    - Current behavior: Coerce`nperseg` to input signal length, warn user, proceed
    - This PR: `ValueError` on `nperseg` being longer than input signal length.
    - I'm not in favor of this change. With a string-like `window`, the values are obtained by a call to `get_window`, which makes it fine to use the coerced `nperseg` and obtain accurate results (albeit from a single FFT segment). I think the current behavior should stand.

- User-specified `nperseg` longer than input signal, string-like `window`
    - Current behavior: Coerce`nperseg` to input signal length, warn user, proceed
    - This PR: `ValueError` inside of `np.lib.stride_tricks.as_strided` in `fft_helper`
    - **Regression**! Line 1000 in `_triage_segments` is too indented. The `nperseg > input_length` check should happen whether or not `nperseg is None`. The code then proceeds with an incorrect `nperseg`. The test suite passes, so a test case must be added that would fail on the current code.

**Comment by NickleDave** [comment]
@e-q thank you for taking the time to go over that. The cases you wrote notes on helped me think this through.

It seems like the issues arise from the last two cases you considered.
* Unspecified ```nperseg```, string-like ```window```, input signal length shorter than 256
  * Current behavior: coerce ```nperseg``` to input signal length, warn user, proceed
    * pros:
      * this is the current behavior--a change could affect some existing code
      * gives accurate results, albeit with a single segment
    * cons (in my view):
      * not clear why it's okay to issue a warning in this case but it's better to raise a ValueError in other cases where the window is longer than the input signal; i.e. why raise the ValueError for an ndarray window that's longer than the input signal but not for a "signal.window" that's longer than the input signal? Why only do so if the input signal is less than the default? Would this still make sense if the default was say 128 or 512?
      * is it really foolproof to assume the user wants a window as long as the input signal? E.g. for a spectrogram this would lose all temporal resolution. I can think of cases where that might matter, for example if the user has few samples (less than 256) but there's long periods of time between each sample
         * so wouldn't it be better to just always ask the user to consciously choose a window by raising a ValueError?
      * Also a warning might slip by unnoticed, e.g., in a stream of talkback from a long-running script

* User-specified ```nperseg``` longer than input signal, string-like ```window```
  * thank you for catching my indentation mistake in line 1000. I will fix that.
  * Do you think this case should raise a ```ValueError```? I mean from line 1000 in ```_triage_segments```, not by mistake from within ```_fft_helper```
  * Again it's not clear to me why this case should raise a ```ValueError``` but the case above shouldn't. In both cases the problem is that the window is longer than the input signal. 
  * If you think this case should raise a ```ValueError```, then once I fix the indentation mistake, the tests will verify the error is raised as expected

**Comment by NickleDave** [comment]
@e-q @Eric89GXL @rgommers any comments?

**Comment by NickleDave** [comment]
Ok I don't want to be that person who acts like their PR is the most important thing in the world, was just trying to knock this off my to-do list.

I modified _triage_segment to maintain the previous behavior where it will only trigger a warning when ```nperseg``` is not specified and the input signal is shorter than the default ```nperseg``` of 256.

But if the user specifies ```nperseg``` and the input signal is shorter than the specified ```nperseg```, that results in a ```ValueError```. @e-q I think that's what you wanted

I saw signal.window will get added to the public API, not sure how that will change spectral.py--just let me know if there's anything else I can do with this

**Comment by Eric89GXL** [comment]
Thanks for taking a look again @NickleDave. I'll let @e-q respond to the specific comments in time. I'll mark this as 0.19 to hopefully get it done in the next week.

**Comment by e-q** [comment]
@NickleDave: The basic idea I have is that with a string-like `window` only raising an error when the given value of `nperseg` doesn't work is that the function that ends up being called to generate the window values generally does so from analytic expressions. In that case, we still have the desired spectral properties of the intended windows even if the value of `nperseg` is changed. This should be the case whenever `nperseg` needs to be shortened, whether from the default value or from a user supplied value.

You are correct that in the case of a call to `spectrogram`, no temporal information will be present. However, it is still possible to produce a correct result, and the user will still receive a warning that their call may not have been what they wanted. This behavior has been around for a while, and I do not recall seeing issues raised about it.

This fundamental difference from the case where the window is supplied by numbers in an array is that we have no guaranteed way of resampling the window function without spoiling the desired spectral qualities. In the absence of the ability to calculate a correct result, we should raise an error.

**Comment by NickleDave** [comment]
@e-q okay, so if I hear what you're saying, the previous behavior should be preserved when users specify a string-like window but the input signal is shorter than the default ```nperseg```. In all such cases, ```_triage_segments``` will issue a ```UserWarning``` and adjust ```nperseg``` to length size of the input signal.

I added a couple more commits so that is the case.

Your arguments for maintaining this behavior are that 
(1) string-like windows correspond to known analytic functions, so the properties of spectra obtained with those functions will be maintained when ```nperseg``` is changed (but we can't make the same guarantee about properties for an arbitrary array-like window)
(2) this behavior has been around for a while and nobody's had an issues with it.

Makes sense to me :)

Please let me know if there's other changes I need to make

**Review by e-q — APPROVED** [review]

**Comment by e-q** [comment]
Ok, this looks good! Thanks for being patient with me :)
I will merge this shortly.

**Comment by e-q** [comment]
Squashed and merged in 93a0ea9

**Comment by NickleDave** [comment]
Thanks for being patient with me! I learned a lot doing this.
